hearth/lib/hearth/interceptor.rb

```diff
@@ -1,66 +1,464 @@
+  #    #=> "Before execution"
+  #    #=> "After execution"
+  #
+  class Interceptor
     # @api private
-    # Apply all interceptors that implement the given hook
-    # @param [Symbol] hook the specific hook to apply for
-    # @param input operation input
-    # @param [Hearth::Context] context
-    # @param [Hearth::Output] output may be nil if unavailable
-    # @param [Boolean] aggregate_errors if true all interceptors are run and
-    #   only the last error is returned.  If false, returns immediately if an
-    #   error is encountered.
-    # @return nil if successful, an exception otherwise
-    def self.apply(hook:, input:, context:, output:, aggregate_errors: false)
-      i_ctx = interceptor_context(input, context, output)
-      last_error = nil
-      context.interceptors.each do |i|
-        next unless i.respond_to?(hook)
-
-        begin
-          i.send(hook, i_ctx)
-        rescue StandardError => e
-          context.logger.error(last_error) if last_error
-          last_error = e
-          break unless aggregate_errors
-        end
-      end
+    @hooks = [
+      READ_BEFORE_EXECUTION = :read_before_execution,
+      MODIFY_BEFORE_SERIALIZATION = :modify_before_serialization,
+      READ_BEFORE_SERIALIZATION = :read_before_serialization,
+      READ_AFTER_SERIALIZATION = :read_after_serialization,
+      MODIFY_BEFORE_RETRY_LOOP = :modify_before_retry_loop,
+      READ_BEFORE_ATTEMPT = :read_before_attempt,
+      MODIFY_BEFORE_ATTEMPT_COMPLETION = :modify_before_attempt_completion,
+      READ_AFTER_ATTEMPT = :read_after_attempt,
+      MODIFY_BEFORE_SIGNING = :modify_before_signing,
+      READ_BEFORE_SIGNING = :read_before_signing,
+      READ_AFTER_SIGNING = :read_after_signing,
+      MODIFY_BEFORE_TRANSMIT = :modify_before_transmit,
+      READ_BEFORE_TRANSMIT = :read_before_transmit,
+      READ_AFTER_TRANSMIT = :read_after_transmit,
+      MODIFY_BEFORE_DESERIALIZATION = :modify_before_deserialization,
+      READ_BEFORE_DESERIALIZATION = :read_before_deserialization,
+      READ_AFTER_DESERIALIZATION = :read_after_deserialization,
+      MODIFY_BEFORE_COMPLETION = :modify_before_completion,
+      READ_AFTER_EXECUTION = :read_after_execution
+    ]
+
+    # @param [Hash<Symbol, Proc>] callbacks
+    #   A hash of hook names to callbacks. The callbacks will be invoked
+    #   when the hook is called. The callback will be passed the
+    #   {InterceptorContext} for the current execution.
+    def initialize(callbacks = {})
+      @callbacks = {}
 
-      set_output_error(last_error, context, output)
+      Interceptor.hooks.each do |hook|
+        next unless callbacks[hook]
 
-      last_error
+        unless valid_callback?(callbacks[hook])
+          raise ArgumentError,
+                "#{hook} must be a callable with arity 1 (context)"
+        end
+
+        @callbacks[hook] = callbacks[hook]
+        define_singleton_method(hook) do |context|
+          @callbacks[hook].call(context)
```

Comment:
I think we can simplify this even more.  We can just pass the callback to the method and then we don't even need to store them at all in `@callbacks`:

```
define_singleton_method(hook, callbacks[hook])
```

Reply:
I'll try this.

Reply:
It works - however - it does not work for capturing input in our poller. I'm not entirely sure on the reason, but it looks to register differently than when I pass the proc as a block - maybe because it's double wrapped?